trie/stacktrie.go

```diff
@@ -17,183 +17,100 @@
-		writeFn:  writeFn,
+		writeFn: writeFn,
+		root:    stPool.Get().(*stNode),
+		h:       newHasher(false),
 	}
 }
 
 // NewStackTrieWithOwner allocates and initializes an empty trie, but with
 // the additional owner field.
 func NewStackTrieWithOwner(writeFn NodeWriteFunc, owner common.Hash) *StackTrie {
-	return &StackTrie{
-		owner:    owner,
-		nodeType: emptyNode,
-		writeFn:  writeFn,
-	}
+	stack := NewStackTrie(writeFn)
+	stack.owner = owner
+	return stack
 }
 
-// NewFromBinary initialises a serialized stacktrie with the given db.
-func NewFromBinary(data []byte, writeFn NodeWriteFunc) (*StackTrie, error) {
-	var st StackTrie
-	if err := st.UnmarshalBinary(data); err != nil {
-		return nil, err
-	}
-	// If a database is used, we need to recursively add it to every child
-	if writeFn != nil {
-		st.setWriter(writeFn)
+// Update inserts a (key, value) pair into the stack trie.
+func (stack *StackTrie) Update(key, value []byte) error {
+	k := keybytesToHex(key)
+	if len(value) == 0 {
+		panic("deletion not supported")
 	}
-	return &st, nil
+	stack.insert(stack.root, k[:len(k)-1], value, nil)
+	return nil
 }
 
-// MarshalBinary implements encoding.BinaryMarshaler
-func (st *StackTrie) MarshalBinary() (data []byte, err error) {
-	var (
-		b bytes.Buffer
-		w = bufio.NewWriter(&b)
-	)
-	if err := gob.NewEncoder(w).Encode(struct {
-		Owner    common.Hash
-		NodeType uint8
-		Val      []byte
-		Key      []byte
-	}{
-		st.owner,
-		st.nodeType,
-		st.val,
-		st.key,
-	}); err != nil {
-		return nil, err
-	}
-	for _, child := range st.children {
-		if child == nil {
-			w.WriteByte(0)
-			continue
-		}
-		w.WriteByte(1)
-		if childData, err := child.MarshalBinary(); err != nil {
-			return nil, err
-		} else {
-			w.Write(childData)
-		}
+// MustUpdate is a wrapper of Update and will omit any encountered error but
+// just print out an error message.
+func (st *StackTrie) MustUpdate(key, value []byte) {
+	if err := st.Update(key, value); err != nil {
+		log.Error("Unhandled trie error in StackTrie.Update", "err", err)
 	}
-	w.Flush()
-	return b.Bytes(), nil
-}
-
-// UnmarshalBinary implements encoding.BinaryUnmarshaler
-func (st *StackTrie) UnmarshalBinary(data []byte) error {
-	r := bytes.NewReader(data)
-	return st.unmarshalBinary(r)
 }
 
-func (st *StackTrie) unmarshalBinary(r io.Reader) error {
-	var dec struct {
-		Owner    common.Hash
-		NodeType uint8
-		Val      []byte
-		Key      []byte
-	}
-	if err := gob.NewDecoder(r).Decode(&dec); err != nil {
-		return err
-	}
-	st.owner = dec.Owner
-	st.nodeType = dec.NodeType
-	st.val = dec.Val
-	st.key = dec.Key
-
-	var hasChild = make([]byte, 1)
-	for i := range st.children {
-		if _, err := r.Read(hasChild); err != nil {
-			return err
-		} else if hasChild[0] == 0 {
-			continue
-		}
-		var child StackTrie
-		if err := child.unmarshalBinary(r); err != nil {
-			return err
-		}
-		st.children[i] = &child
-	}
-	return nil
+func (stack *StackTrie) Reset() {
+	stack.owner = (common.Hash{})
```

Comment:
> the owner and writeFn unchanged.

I think that is dangerous. The idea of `Reset` is that it can be called before entering it into the pool, where the object may live indefinitely. So we need to clear out anything which may hold memory references, and IMO the `writeFn` is very much a likely culprit to have references to local variables in local scopes. 

The `owner` we might do away with completely, as you pointed out below.

Reply:
Well, regardless: an object that is newly `Reset` should not hold external references, IMO